genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/S3ArchivalServiceImpl.java

```diff
@@ -0,0 +1,172 @@
+import com.netflix.genie.agent.execution.services.ArchivalService;
+import com.netflix.genie.common.internal.jobs.JobConstants;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import javax.validation.constraints.NotNull;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.net.URI;
+
+/**
+ * Implementation of ArchivalService for S3.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Component
+@Lazy
+@Slf4j
+class S3ArchivalServiceImpl implements ArchivalService {
+
+    private final AmazonS3 amazonS3;
+
+    S3ArchivalServiceImpl(final AmazonS3 amazonS3) {
+        this.amazonS3 = amazonS3;
+    }
+
+    /**
+     * Store the contents of a file to S3.
+     * For a directory all its contents are recursively traversed. Each of the file is
+     * stored as a an object in S3. Each empty directory is also stored as an object with
+     * empty content.
+     * <p>
+     * Example 1
+     * <p>
+     * For a dir hierarchy
+     * /tmp/${jobid}/genie/applications (empty dir)
+     * /tmp/${jobid}/genie/run (file)
+     * /tmp/${jobid}/genie/commands/presto (file)
+     * <p>
+     * dir to be archived - /tmp/${jobid}
+     * target location - s3://$bucketName/{$jobid}
+     * <p>
+     * following structure is created in S3
+     * s3://$bucketName/{$jobid}/genie/applications/ -> empty content
+     * s3://$bucketName/{$jobid}/genie/run -> contents of run file
+     * s3://$bucketName/{$jobid}/genie/commands/presto -> contents of presto file
+     * <p>
+     * Example 2
+     * <p>
+     * file to archive - /tmp/run
+     * target location - s3://$bucketName/run
+     * <p>
+     * following structure is created in S3
+     * s3://$bucketName/abc -> contents of run file
+     *
+     * @param file              file/directory to archive.
+     * @param targetLocationURI target location uri for the file
+     * @throws ArchivalException if archival fails
+     */
+
+    @Override
+    public void archive(
+        @NotNull final File file,
+        @NotNull final URI targetLocationURI
+    ) throws ArchivalException {
+        log.info("Archiving to location: {}", targetLocationURI);
+
+        try {
+            final AmazonS3URI targetLocationAmazonS3URI = new AmazonS3URI(targetLocationURI);
+            archiveRecursively(file, targetLocationAmazonS3URI);
+        } catch (Exception e) {
+            log.info("Error archiving to target location: {} ", targetLocationURI);
+            throw new ArchivalException("Error archiving file: " + file.getName(), e);
+        }
+    }
+
+    /**
+     * Traverse a directory recursively and archive the files and directories
+     * underneath it. If the directory is empty, create an empty object in S3.
+     * For a regular child file create an object in S3 and copy the file contents.
+     * For a empty child directory create an empty object in S3
```

Comment:
Why bother creating an empty object in S3 for an empty directory?

Reply:
Allow to list vs. throw 404.

Reply:
For backward compatibility. Today on the v3 api when one is browsing the output folder, empty directories are shown. If we don't store empty directories in S3 they cannot be shown.

Reply:
Ok, will remove empty folder archival then.

Reply:
This is a fair point. Anecdotally it feels to me like this is useless/wasteful I can't definitively say it will be that way in every possible case. For now we could leave it in and come back later if we decide it really isn't worth it or is wasting too much time. I'm not sure I can think of a case where an empty folder is useful information for any command/process but perhaps there's something you all can think of that i'm missing. Maybe the absence of an expected file is useful information though I'd argue that the directory not being present clearly also indicates the expected file doesn't exist.